template/template.go

```diff
@@ -1,46 +1,124 @@
 package template
 
 import (
+	"strings"
+
 	"github.com/hashicorp/hil"
 	"github.com/hashicorp/hil/ast"
+	"github.com/openpixel/rise/config"
 	"github.com/openpixel/rise/interpolation"
 )
 
 // Template is a container for holding onto the ast Variables
 type Template struct {
-	vars map[string]ast.Variable
+	vars      map[string]ast.Variable
+	templates map[string]ast.Variable
 }
 
 // NewTemplate will prepare a template object for use
-func NewTemplate(vars map[string]ast.Variable) (*Template, error) {
-	if vars == nil {
-		vars = make(map[string]ast.Variable)
-	}
+func NewTemplate(configResult *config.Result) (*Template, error) {
 	return &Template{
-		vars: vars,
+		vars:      configResult.Variables,
+		templates: configResult.Templates,
 	}, nil
 }
 
 func (t *Template) buildConfig() *hil.EvalConfig {
+	vars := make(map[string]ast.Variable)
+
+	for k, v := range t.vars {
+		vars[k] = v
+	}
+	for k, v := range t.templates {
+		vars[k] = v
+	}
+
 	return &hil.EvalConfig{
 		GlobalScope: &ast.BasicScope{
 			FuncMap: interpolation.CoreFunctions,
-			VarMap:  t.vars,
+			VarMap:  vars,
 		},
 	}
 }
 
 // Render will parse the provided text and interpolate the known variables/functions
 func (t *Template) Render(text string) (hil.EvaluationResult, error) {
+	var resultErr error
+
+	config := t.buildConfig()
+
 	tree, err := hil.Parse(text)
 	if err != nil {
 		return hil.InvalidResult, err
 	}
 
-	result, err := hil.Eval(tree, t.buildConfig())
+	fn := func(n ast.Node) ast.Node {
+		if resultErr != nil {
+			return n
+		}
+
+		var varName string
+
+		switch vn := n.(type) {
+		case *ast.VariableAccess:
+			varName = vn.Name
+		case *ast.Index:
+			if va, ok := vn.Target.(*ast.VariableAccess); ok {
+				varName = va.Name
+			}
+			if va, ok := vn.Key.(*ast.VariableAccess); ok {
+				varName = va.Name
+			}
+		default:
+			return n
+		}
+
+		if strings.HasPrefix(varName, "tmpl.") {
```

Comment:
I wonder if I was to perform interpolation on each variable (by looking at prefix `var.` and then run the visit function on all new template nodes, if this would increase performance. At this point, I am visiting the tree just for templates and then running Eval on the tree, which visits the nodes again.